jwt/audience.go

```diff
@@ -3,78 +3,71 @@ package jwt
 import (
 	"errors"
 	"fmt"
-	"net"
-	"net/url"
+	"strconv"
 	"strings"
 )
 
-// Audience must be the base URL from the request including protocol, domain,
-// and port if applicable for the domains you specify in your IAP proxy. For
-// example, https://example.com or https://foo.example.com:port.
-type Audience url.URL
+// From the IAP docs at https://cloud.google.com/iap/docs/signed-headers-howto:
+// Audience must be a string with the following values:
+// * App Engine: /projects/PROJECT_NUMBER/apps/PROJECT_ID
+// * Compute Engine and Container Engine: /projects/PROJECT_NUMBER/global/backendServices/SERVICE_ID
+type Audience string
 
-// NewAudience returns an Audience from a URL.
-func NewAudience(u *url.URL) *Audience {
-	return (*Audience)(u)
-}
-
-// Sanitize normalizes the structure of the Audience's URL and validates it.
-func (aud *Audience) Sanitize() error {
-	aud.Scheme = strings.ToLower(aud.Scheme)
-	var err error
-	var port string
-	host := aud.Host
-	if strings.LastIndex(host, ":") >= 0 {
-		host, port, err = net.SplitHostPort(host)
-		if err != nil {
-			return err
-		}
-	}
-	if len(port) == 0 {
-		defaultPort := "80"
-		if aud.Scheme == "https" {
-			defaultPort = "443"
-		}
-		aud.Host = net.JoinHostPort(host, defaultPort)
-	}
-	return aud.Validate()
+// NewAudience returns an Audience from a string.
+func NewAudience(u string) *Audience {
+	aud := Audience(u)
+	return &aud
 }
 
 // Validate performs error checking on the Audience's URL.
 func (aud *Audience) Validate() error {
-	if aud.Scheme != "http" && aud.Scheme != "https" {
-		return fmt.Errorf("Unexpected scheme: %s", aud.Scheme)
+	rawAudience := string(*aud)
+	p := strings.SplitN(rawAudience, "/", 5)
+	if len(p) != 5 {
+		return errors.New("invalid audence length")
 	}
-	host, _, err := net.SplitHostPort(aud.Host)
-	if err != nil {
-		return err
-	}
-	if len(host) == 0 {
-		return errors.New("Host not specified")
-	}
-	if aud.User != nil {
-		return errors.New("Not expecting user")
-	}
-	if len(aud.Path) != 0 || len(aud.RawPath) != 0 {
-		return errors.New("Not expecting path")
+
+	if p[1] != "projects" {
+		return errors.New("audience must start with a project")
 	}
-	if len(aud.RawQuery) != 0 {
-		return errors.New("Not expecting query")
+
+	var (
+		projectNumber = p[2]
+		audienceType  = p[3]
+		audienceID    = p[4]
+	)
+
+	if _, err := strconv.ParseUint(projectNumber, 10, 64); err != nil {
```

Comment:
I would probably leave this check out. They can be 64bit or maybe bigint, who knows. It may change in the future too. I would probably just check if it's empty or not.